truss-chains/truss_chains/streaming.py

```diff
@@ -0,0 +1,380 @@
+            raise ValueError("Stream does not contain header.")
+        return self._stream_types.header_t.model_validate_json(data_bytes)
+
+
+class _FooterReadMixin(_Streamer[ItemT, HeaderTT, FooterT]):
+    _footer_data: Optional[bytes]
+
+    async def read_footer(
+        self: _StreamReaderProtocol[ItemT, HeaderTT, FooterT],
+    ) -> FooterT:
+        if self._footer_data is None:
+            delimiter, data_bytes = await self._read()
+            if delimiter != _Delimiter.FOOTER:
+                raise ValueError("Stream does not contain footer.")
+            self._footer_data = data_bytes
+
+        footer = self._stream_types.footer_t.model_validate_json(self._footer_data)
+        self._footer_data = None
+        return footer
+
+
+class StreamReaderWithHeader(
+    _StreamReader[ItemT, HeaderT, FooterTT], _HeaderReadMixin[ItemT, HeaderT, FooterTT]
+): ...
+
+
+class StreamReaderWithFooter(
+    _StreamReader[ItemT, HeaderTT, FooterT], _FooterReadMixin[ItemT, HeaderTT, FooterT]
+): ...
+
+
+class StreamReaderFull(
+    _StreamReader[ItemT, HeaderT, FooterT],
+    _HeaderReadMixin[ItemT, HeaderT, FooterT],
+    _FooterReadMixin[ItemT, HeaderT, FooterT],
+): ...
+
+
+@overload
+def stream_reader(
+    types: StreamTypes[ItemT, None, None],
+    stream: AsyncIterator[bytes],
+) -> _StreamReader[ItemT, None, None]: ...
+
+
+@overload
+def stream_reader(
+    types: StreamTypes[ItemT, HeaderT, None],
+    stream: AsyncIterator[bytes],
+) -> StreamReaderWithHeader[ItemT, HeaderT, None]: ...
+
+
+@overload
+def stream_reader(
+    types: StreamTypes[ItemT, None, FooterT],
+    stream: AsyncIterator[bytes],
+) -> StreamReaderWithFooter[ItemT, None, FooterT]: ...
+
+
+@overload
+def stream_reader(
+    types: StreamTypes[ItemT, HeaderT, FooterT],
+    stream: AsyncIterator[bytes],
+) -> StreamReaderFull[ItemT, HeaderT, FooterT]: ...
+
+
+def stream_reader(
+    types: StreamTypes[ItemT, HeaderTT, FooterTT],
+    stream: AsyncIterator[bytes],
+) -> _StreamReader:
+    if types.header_t is None and types.footer_t is None:
+        return _StreamReader(types, stream)
+    if types.header_t is None:
+        return StreamReaderWithFooter(types, stream)
+    if types.footer_t is None:
+        return StreamReaderWithHeader(types, stream)
+
+    return StreamReaderFull(types, stream)
+
+
+# Writing ##############################################################################
+
+
+class _StreamWriterProtocol(Protocol[ItemT, HeaderTT, FooterTT]):
+    _stream_types: StreamTypes[ItemT, HeaderTT, FooterTT]
+
+    def _serialize(self, obj: pydantic.BaseModel, delimiter: _Delimiter) -> bytes: ...
+
+
+class _StreamWriter(_Streamer[ItemT, HeaderTT, FooterTT]):
+    @staticmethod
+    def _pack_tag(delimiter: _Delimiter, length: int) -> bytes:
+        return struct.pack(">BI", delimiter.value, length)
+
+    def _serialize(self, obj: pydantic.BaseModel, delimiter: _Delimiter) -> bytes:
+        data_bytes = obj.model_dump_json().encode()
```

Comment:
just to be clear, if someone has binary data (say audio) that they are sending back, they don't have to use this right? They would just stream the bytes directly?

I'm not sure how a pydantic model would serialize a field that contained binary data

Reply:
If you want to return raw binary data, you don't need any of these streamers, you can directly stream the bytes.